### PR TITLE
[Fix] Honor FlashMLA natural-log LSE in DCP reduction

### LINKED ISSUE #33064: [Bug] FlashMLA DCP combines natural-log LSE with base-2 correction
## Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] This report includes environment information and a minimal reproducer.
- [x] This is a bug report rather than a general question.
- [x] This report is in English.

## Describe the bug

FlashMLA returns softmax LSE in the natural-log domain, but the MLA decode DCP reduction currently combines it as if it were base-2. This produces incorrect cross-rank softmax weights and therefore incorrect attention output whenever FlashMLA is used with decode context parallelism.

There are two affected communication paths:

- The all-gather/reduce-scatter path calls `correct_attn_out`, whose Triton kernel unconditionally uses `tl.exp2` and `tl.log2`.
- The A2A path calls `dcp_a2a_lse_reduce(..., is_lse_base_on_e=False)` for both FlashInfer MLA and FlashMLA.

This is not only a numerical-stability or performance issue. It changes the model output. FlashMLA's `flash_mla_with_kvcache` returns `softmax_lse`, and its AOT tests validate that result against `torch.logsumexp`, i.e. natural-log LSE. FlashInfer MLA returns base-2 LSE, so the reduction must select the log base by attention backend.

For a minimal numerical example, let two DCP shards have partition functions `Z = [2, 8]`, and let the local attention output on the second shard be `10`. The correct cross-shard weight is `8 / (2 + 8)`, so the corrected local contribution is `8`. Treating `[ln(2), ln(8)]` as base-2 LSE instead gives:

```python
import math

wrong = 10 * (2 ** math.log(8)) / (
    2 ** math.log(2) + 2 ** math.log(8)
)
print(wrong)  # 7.2330317031552385; expected 8.0
```

## Reproduction

The kernel-level regression below reproduces the affected all-gather/reduce-scatter correction directly. It uses the natural-log LSE values emitted by FlashMLA:

```python
import torch

from sglang.kernels.ops.attention.dcp_kernels import correct_attn_out

local_output = torch.tensor([[[10.0]]], device="cuda", dtype=torch.bfloat16)
lses = torch.log(torch.tensor([[[2.0]], [[8.0]]], device="cuda"))

corrected, _ = correct_attn_out(
    local_output,
    lses,
    cp_rank=1,
    ctx=None,
    new_output=torch.empty((1, 1, 1), device="cuda"),
    is_lse_base_on_e=True,
)

torch.testing.assert_close(
    corrected.cpu(), torch.tensor([[[8.0]]]), atol=1e-5, rtol=1e-5
)
```

The proposed fix adds an explicit LSE-base flag to the AG/RS correction kernel and selects natural-log mode only when `current_attention_backend == "flashmla"`. The default remains base-2, preserving FlashInfer MLA behavior. The existing A2A reducer already supports this flag, so its call site is corrected as well.

Validation on one NVIDIA H20 GPU:

```text
python3 -m pytest -q test/registered/kernels/test_dcp_lse_combine.py \
  -k "flashmla_natural_log_lse_correction or flashmla_selects_natural_log_lse"

2 passed, 21 deselected in 24.83s

python3 -m pytest -q test/registered/kernels/test_dcp_lse_combine.py

23 passed in 25.88s
```

## Environment

- Base revision: `sgl-project/sglang@585a7d05e384c79ee17bf3185c7469fe8ce75946` (latest `main` when reproduced)
- GPU: NVIDIA H20 (SM90, 96 GB)
- NVIDIA driver: 580.105.08
- Runtime used for the GPU regression: SGLang v0.5.16-based CUDA container with the branch source mounted through `PYTHONPATH`

`python3 -m sglang.check_env` output:

```text
Python: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H20
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 580.105.08
PyTorch: 2.11.0+cu130
sglang: 0.5.15.post2.dev1418+g907d05c86
sglang-kernel: 0.4.5
flashinfer_python: 0.6.15.post1
flashinfer_cubin: 0.6.15.post1
flashinfer_jit_cache: 0.6.15.post1+cu130
triton: 3.6.0
transformers: 5.12.1
torchao: 0.17.0+cu130
numpy: 2.3.5
aiohttp: 3.14.3
fastapi: 0.140.13
huggingface_hub: 1.25.1
interegular: 0.3.3
modelscope: 1.39.0
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.2
psutil: 7.2.2
pydantic: 2.13.4
python-multipart: 0.0.32
pyzmq: 27.1.0
uvicorn: 0.51.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.13.0
anthropic: 0.120.2
litellm: Module Not Found
torchcodec: 0.11.1+cu130

NVIDIA Topology:
        GPU0 GPU1 GPU2 GPU3 GPU4 GPU5 GPU6 GPU7
GPU0     X   NV18 NV18 NV18 NV18 NV18 NV18 NV18
GPU1    NV18  X   NV18 NV18 NV18 NV18 NV18 NV18
GPU2    NV18 NV18  X   NV18 NV18 NV18 NV18 NV18
GPU3    NV18 NV18 NV18  X   NV18 NV18 NV18 NV18
GPU4    NV18 NV18 NV18 NV18  X   NV18 NV18 NV18
GPU5    NV18 NV18 NV18 NV18 NV18  X   NV18 NV18
GPU6    NV18 NV18 NV18 NV18 NV18 NV18  X   NV18
GPU7    NV18 NV18 NV18 NV18 NV18 NV18 NV18  X

ulimit soft: 1024
```

### PR DESCRIPTION
**Pull request by CyberSecurityErial**
## Motivation

FlashMLA returns natural-log softmax LSE, while the MLA DCP merge paths currently treat all decode LSE as base-2. As a result, FlashMLA + DCP computes incorrect cross-rank softmax weights and attention output.

For partition functions `Z = [2, 8]` and a second-shard local output of `10`, the correct contribution is `8`. Interpreting `[ln(2), ln(8)]` with `exp2` produces `7.2330317` instead.

Fixes #33064.

## Modifications

- Make the AG/RS Triton correction kernel select `exp/log` or `exp2/log2` through a compile-time LSE-base flag.
- Propagate that flag through `cp_lse_ag_out_rs_mla`.
- Select natural-log LSE for the `flashmla` attention backend in both MLA DCP communication paths:
  - all-gather LSE + reduce-scatter output;
  - A2A/FlashInfer-A2A LSE reduction.
- Keep base-2 as the default, so FlashInfer MLA behavior is unchanged.
- Add a GPU regression for the exact natural-log weighting error and a routing test for backend selection.

## Accuracy Tests

Tested on one NVIDIA H20 GPU with the branch source based on `585a7d05e384c79ee17bf3185c7469fe8ce75946`:

```text
PYTHONPATH=/work/flashmla-lse/python python3 -m pytest -q \
  test/registered/kernels/test_dcp_lse_combine.py \
  -k "flashmla_natural_log_lse_correction or flashmla_selects_natural_log_lse"

2 passed, 21 deselected in 24.83s
```

The full related test file also passes:

```text
PYTHONPATH=/work/flashmla-lse/python python3 -m pytest -q \
  test/registered/kernels/test_dcp_lse_combine.py

23 passed in 25.88s
```

The warnings emitted by the container are unrelated environment/deprecation warnings; there were no test failures.

## Speed Tests and Profiling

Not applicable. This fixes output correctness. The log-base branch is a Triton `tl.constexpr`, so it is specialized at compilation rather than selected per element at runtime.

## Checklist

- [x] Format the modified files according to the project formatter.
- [x] Add unit tests.
- [ ] Update documentation (not needed: no public API or user-facing configuration change).
- [x] Provide accuracy test results.
- [x] Follow the SGLang code style guidance.

---
### CI States

Latest PR Test (Base): :x: [Run #30867184213](https://github.com/sgl-project/sglang/actions/runs/30867184213)
Latest PR Test (Extra): :x: [Run #30867184031](https://github.com/sgl-project/sglang/actions/runs/30867184031)

### CODE REVIEW

**Review by kpham-sgl — APPROVED** [review]
LGTM. Can you resolve conflicts

**Comment by CyberSecurityErial** [comment]
Resolved the conflicts by rebasing onto the current `main` (`3960983753`). The new head is `b3a816b2c9`.

The only conflict resolution was adopting the upstream configuration accessor, `get_parallel().dcp_comm_backend`, while preserving the natural-log LSE flag propagation through both DCP reduction paths. Post-rebase validation completed:

- `git diff --check`
- `py_compile` for all four touched Python files
- `range-diff` against the approved patch; no other semantic drift
- the same functional patch previously passed the full H20 regression file: `23 passed`

I did not rerun GPU tests for the rebased SHA because the available H20 node is currently serving active workloads. The current red checks are gate-only (`Missing required label run-ci`); all actual test suites were skipped.

@kpham-sgl, could you please re-approve the rebased head and add the `run-ci` label / rerun CI? Thanks.

**Comment by CyberSecurityErial** [comment]
&nbsp;Hi Khoa,


Thanks for the review. I’ve rebased the PR and resolved the conflicts. The PR is now mergeable.


Could you please re-approve it and trigger CI? The current checks are blocked by the missing run-ci label.


Best,
Chutian Wang

------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "sgl-project/sglang"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2026年8月4日(星期二) 凌晨2:48
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [sgl-project/sglang] [Fix] Honor FlashMLA natural-log LSE in DCP reduction (PR #33065)





 
@kpham-sgl approved this pull request.
 
LGTM. Can you resolve conflicts
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!

**Comment by CyberSecurityErial** [comment]
Update: the H20 node became available, so I reran the complete regression file on the rebased head `b3a816b2c9`:

```text
23 passed, 15 warnings in 14.07s
exit code: 0
```

This confirms the rebased patch passes the same H20 regression.

**Comment by CyberSecurityErial** [comment]
The CI passed all 23 tests, and the PR is now mergeable.&nbsp;
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "sgl-project/sglang"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2026年8月4日(星期二) 上午9:30
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [sgl-project/sglang] [Fix] Honor FlashMLA natural-log LSE in DCP reduction (PR #33065)



kpham-sgl left a comment (sgl-project/sglang#33065)
 
/rerun-test test/registered/kernels/test_dcp_lse_combine.py
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!
